File: python/149. Max Points on a Line.py
# Definition for a point.
# class Point(object):
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b
import sys
class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        if len(points) <= 2:
            return len(points)
        result = 0
        for i in range(len(points)):
            d = {}
            overlap, curmax = 0, 0
            for j in range(i + 1, len(points)):
                dx, dy = points[j].x - points[i].x, points[j].y - points[i].y
                if dx == 0 and dy == 0:
                    overlap += 1
                    continue
                slope = dy * 1.0 / dx if dx != 0 else sys.maxint
                if slope in d:
                    d[slope] += 1
                else:
                    d[slope] = 1
                curmax = max(curmax, d[slope])
            result = max(result, curmax + overlap + 1)
        return result
                
                
                
                
                

File: cpp/101-150/149. Max Points on a Line.cc
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
private:
    int gcd(int a, int b) {
        while (b) {
            int tmp = b;
            b = a % b;
            a = tmp;
        }
        return a;
    }
public:
    int maxPoints(vector<Point>& points) {
        if (points.empty())
            return 0;
        if (points.size() == 1)
            return 1;
        int res = 0;
        for (int i = 0; i < points.size(); i++) {
            int overlap = 0, curmax = 0;
            unordered_map<pair<int, int>, int> k_count;
            for (int j = i + 1; j < points.size(); j++) {
                int dx = points[j].x - points[i].x;
                int dy = points[j].y - points[i].y;
                if (dx == 0 && dy == 0) {
                    overlap++;
                    continue;
                }
                int dvs = gcd(dx, dy);
                pair<int, int> k = make_pair(dx / dvs, dy / dvs);
                if (k_count.find(k) == k_count.end()) {
                    k_count[k] = 1;
                }
                else
                    k_count[k]++;
                curmax = max(curmax, k_count[k]);
            }
            res = max(res, curmax + overlap + 1);
        }
        return res;
    }
};